nixos/lib/testing-python.nix

```diff
@@ -132,7 +132,7 @@ rec {
 
       invalidNodeNames = lib.filter
         (node: builtins.match "^[A-z_]([A-z0-9_]+)?$" node == null)
-        (builtins.attrNames nodes);
```

Comment:
The old implementation of this check actually looks to be correct. The attribute names should be valid python identifiers, but the hostnames don't have to be. Having two kinds of names is confusing and I wouldn't be surprised if they get mixed up in one or two places.

Reply:
At first I thought that, too. But then I realized it's even worse: the testing framework effectively forbids using RFC1035 host names that contain a "`-`", as they would then be invalid python identifiers.

The situation is dire, but this is "fix" at least is explicit about it in a warning. I hope the following pointers make it easier to peak through the "chaos" (I'm currently in a slow sprint of cleaning this up a bit together with tfc).

- [the python identiers](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/test-driver/test-driver.py#L1003)
- [machine creation](https://github.com/NixOS/nixpkgs/blob/d54f3bba7c2153a879753c733f641c85956095f7/nixos/lib/test-driver/test-driver.py#L999)
- [machine name inferrence from the script name](https://github.com/NixOS/nixpkgs/blob/d54f3bba7c2153a879753c733f641c85956095f7/nixos/lib/test-driver/test-driver.py#L255-L259)
- [the script name](https://github.com/NixOS/nixpkgs/blob/d54f3bba7c2153a879753c733f641c85956095f7/nixos/modules/virtualisation/qemu-vm.nix#L730) using `config.system.name` from which the hostname is inferred

THis needs independent cleanup and I don't think it's tolerable to barring hostnames from using `-` if they want to enjoy testing, so I think we need to find a different solution.

I'll take that into special account during my lceanup.

Reply:
In a private quick back and forth on matrix we concluded the following:

- this is a cover up all the way down
- we need to better expose the [implementation error](https://github.com/NixOS/nixpkgs/compare/dbafcfca4feda411465e73921c6c43e67a60f194..169004c4868fae002c6865025e08838b201a8af7) and eventually [fix it](https://github.com/NixOS/nixpkgs/compare/169004c4868fae002c6865025e08838b201a8af7..dc060ff8b2abef52a75fede1b9ea6d35e280931f)